kymatio/scattering2d/frontend/torch_f.py

```diff
@@ -0,0 +1,178 @@
+            spatial support (M, N) of the input
+        L : int, optional
+            number of angles used for the wavelet transform
+        max_order : int, optional
+            The maximum order of scattering coefficients to compute.
+            Must be either equal to `1` or `2`. Defaults to `2`.
+        pre_pad : boolean
+            controls the padding
+        Psi : dictionary
+            containing the wavelets filters at all resolutions. See
+            filter_bank.filter_bank for an exact description.
+        Phi : dictionary
+            containing the low-pass filters at all resolutions. See
+            filter_bank.filter_bank for an exact description.
+        M_padded, N_padded : int
+             spatial support of the padded input
+
+        Notes
+        -----
+        The design of the filters is optimized for the value L = 8
+
+        pre_pad is particularly useful when doing crops of a bigger
+         image because the padding is then extremely accurate. Defaults
+         to False.
+
+        """
+    def __init__(self, J, shape, L=8, max_order=2, pre_pad=False):
+        super(Scattering2D, self).__init__(J, shape, max_order = max_order)
+        self.L, self.J, self.max_order = L, J, max_order
+        self.pre_pad = pre_pad
+        self.shape = shape
+        if 2 ** J > shape[0] or 2 ** J > shape[1]:
+            raise (RuntimeError('The smallest dimension should be larger than 2^J'))
+        self.build()
+
+    def build(self):
+        self.M, self.N = self.shape
+        self.modulus = Modulus()
+        self.M_padded, self.N_padded = compute_padding(self.M, self.N, self.J)
+        # pads equally on a given side if the amount of padding to add is an even number of pixels, otherwise it adds an extra pixel
+        self.pad = Pad([(self.N_padded - self.N) // 2, (self.N_padded - self.N + 1) // 2, (self.M_padded - self.M) // 2,
+                       (self.M_padded - self.M + 1) // 2], [self.N, self.M], pre_pad=self.pre_pad)
+        self.subsample_fourier = SubsampleFourier()
+        self.filters = self.create_and_register_filters()
+
+    def create_and_register_filters(self):
+        # Create the filters
+        filters = filter_bank(self.M_padded, self.N_padded, self.J, self.L)
+        n = 0
+        self.phi, self.psi = filters['phi'], filters['psi']
+        for c, phi in self.phi.items():
+            if isinstance(c, int):
+                self.phi[c] = torch.from_numpy(self.phi[c])
+                self.register_buffer('tensor' + str(n), self.phi[c])
+                n += 1
+
+        for j in range(len(self.psi)):
+            for k, v in self.psi[j].items():
+                if isinstance(k, int):
+                    self.psi[j][k] = torch.from_numpy(v)
+                    self.register_buffer('tensor' + str(n), self.psi[j][k])
+                    n += 1
+
+    def scattering(self, input):
+        # each time scattering is run, one needs to make sure self.psi and self.phi point to
```

Comment:
Why can't this be done once and for all? Do the buffers move around? If so, what's to guarantee that they stay the same during the call to `scattering2d` below?

Reply:
They do move around. The array buffer is an array of "pointer". Here is an url to an example where it is nicely used: https://pytorch.org/docs/stable/_modules/torch/nn/modules/batchnorm.html (for BNs) I can't think of a simpler way to use it because we need an array/dictionary and the buffers do not seem able to track each tensors of a dictionary. Thus currently, I artifically expand it and track it..

The only guarantee is that the registered pointer will point to the current buffer. An example of failure is when one does `.cuda()`, then the memory changes..

Reply:
So why do we have to fetch them during the `scattering()` call? When we register them, we get the pointers to the registered buffers and assign them to our filters already.

Reply:
at the moment we do a .cuda(), the location changes - it's in order to guarantee that the pointer does pooint to the correct memory location.